vulture/config.py

```diff
@@ -0,0 +1,212 @@
+"""
+This module handles retrieval of configuration values from either the
+command-line arguments or the pyproject.toml file.
+"""
+import argparse
+import sys
+from os.path import abspath, exists
+
+import toml
+
+from .version import __version__
+
+#: Possible configuration options and their respective defaults
+DEFAULTS = {
+    "min_confidence": 0,
+    "paths": [],
+    "exclude": [],
+    "ignore_decorators": [],
+    "ignore_names": [],
+    "make_whitelist": False,
+    "sort_by_size": False,
+    "verbose": False,
+}
+
+#: sentinel value to distinguish between "False" and "no default given"
+MISSING = object()
+
+
+def _check_input_config(data):
+    """
+    Checks the types of the values in *data* against the expected types of
+    config-values. If a value is of the wrong type it will raise a SystemExit.
+    """
+    for key, value in data.items():
+        if key not in DEFAULTS:
+            print(f"Unknown configuration key: {key}", file=sys.stderr)
+            sys.exit("Invalid config")
+        if value is not MISSING and not type(value) is type(  # noqa: E721
+            DEFAULTS[key]
+        ):
+            print(
+                f"Data type for {key} must be {DEFAULTS[key]}", file=sys.stderr
+            )
+            sys.exit("Invalid config")
+
+
+def from_dict(data):
+    """
+    Create a new config dictionary from an existing one, assign possible
+    defaults and warn about unprocessed options.
+    """
+    _check_input_config(data)
+
+    unknown_keys = set(data) - set(DEFAULTS)
+    if unknown_keys:
+        print(
+            f"Unknown configuration keys: {sorted(unknown_keys)}",
+            file=sys.stderr,
+        )
+        sys.exit("invalid Config")
```

Comment:
According to the Python docs sys.exit prints to stderr, so it seems the three calls above can be simplified: https://docs.python.org/3/library/sys.html#sys.exit

Reply:
I had to resort to `prints` as using `sys.stderr` breaks the unit-tests with the `capsys` fixture. When using `sys.exit(str)`, both `stdout` and `stderr` values from `capsys` are empty.

I think it's a valid test to ensure we see something helpful on the console output in this test and would prefer to keep the prints. But it's up to you, if you want me to remove them, I can, and it will still print something to `stderr`, but it won't be testable.

Reply:
I don't think we need to test the output.

Reply:
I now removed the print lines and am now confronted with a failing unit-test which fails and I don't know why. As the print statements are gone, it seems that even pytest is unable to see any output now.

A *personal* opinion I have held back until now is that I generally really dislike seeing `sys.exit` lines. I strongly prefer return values and just one `sys.exit` in the application entry-point. `sys.exit` often kneecaps instrumentation tools and makes debugging harder.

Additionally, I prefer to uses error-codes from [`errno.h`](https://python.readthedocs.io/en/latest/library/errno.html). But this is something I see so rarely done that I'm not hung up on this. Considering that it is used so rarely, it's not a very reliable listing anyway. Even Python itself does not return `ENOENT` on a `FileNotFoundError`.

In any case, removing `sys.exit` at this point would imply too many code-changes. If it is removed you suddenly have to think about return-values which will certainly break many things. And those return values (including exceptions) would bubble up and need to be handled somewhere appropriately.

I don't think this change is worth the effort (at least not in this PR). I will keep digging through the code and try to find out why the test fails.